server/queues/processors/WebhookProcessor.ts

```diff
@@ -0,0 +1,383 @@
+import Logger from "@server/logging/Logger";
+import {
+  User,
+  WebhookDelivery,
+  WebhookSubscription,
+  Document,
+  FileOperation,
+  Collection,
+  Group,
+  Integration,
+  Team,
+  Pin,
+  Star,
+} from "@server/models";
+import {
+  presentCollection,
+  presentDocument,
+  presentFileOperation,
+  presentGroup,
+  presentIntegration,
+  presentPin,
+  presentStar,
+  presentTeam,
+  presentUser,
+  presentWebhook,
+  presentWebhookSubscription,
+} from "@server/presenters";
+import {
+  CollectionEvent,
+  DocumentEvent,
+  Event,
+  FileOperationEvent,
+  GroupEvent,
+  IntegrationEvent,
+  PinEvent,
+  RevisionEvent,
+  StarEvent,
+  TeamEvent,
+  UserEvent,
+  WebhookSubscriptionEvent,
+} from "@server/types";
+import BaseProcessor from "./BaseProcessor";
+
+function assertUnreachable(_: never): never {
+  throw new Error("Didn't expect to get here");
+}
+
+export default class WebhookProcessor extends BaseProcessor {
+  static applicableEvents: ["*"] = ["*"];
+
+  async perform(event: Event) {
+    const webhookSubscriptions = await WebhookSubscription.findAll({
+      where: {
+        enabled: true,
+        teamId: event.teamId,
+      },
+    });
+
+    await Promise.all(
```

Comment:
I think we need to split this processor into two. One part that finds all the applicable subscriptions as you have done here and then just schedules a `DeliverWebhookTask` (or similar) for each. There are a couple of things this achieves…

- If there is an error as part of one webhook delivery and this job retries then it won't cause a duplicate delivery for another webhook in the team that succeeded.
- We can lean on the existing queue retry mechanism to retry individual webhooks
- Spreads the load across multiple machines (there are many workers in the cloud hosted instance) and hopefully results in the http requests overall happening faster

Reply:
Ahh! Ya makes a ton of sense to split the deliveries into separate jobs

Will do!